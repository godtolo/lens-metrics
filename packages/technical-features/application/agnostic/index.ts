export { applicationFeature } from "./src/feature";

export { onLoadOfApplicationInjectionToken } from "./src/start-application/timeslots/on-load-of-application-injection-token";
export { beforeApplicationIsLoadingInjectionToken } from "./src/start-application/timeslots/before-application-is-loading-injection-token";
export { afterApplicationIsLoadedInjectionToken } from "./src/start-application/timeslots/after-application-is-loaded-injection-token";

export type { StartApplication } from "./src/start-application/start-application.injectable";
export { startApplicationInjectionToken } from "./src/start-application/start-application.injectable";

export { applicationInformationToken } from "./src/application-information-token";
export type { ApplicationInformation } from "./src/application-information-token";
