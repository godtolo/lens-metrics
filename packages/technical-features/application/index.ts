export { applicationInformationToken } from "./src/application-information-token";
export type { ApplicationInformation } from "./src/application-information-token";

