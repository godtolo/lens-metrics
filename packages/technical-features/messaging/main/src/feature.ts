import { autoRegister } from "@ogre-tools/injectable-extension-for-auto-registration";
import { getFeature } from "@k8slens/feature-core";

export const messagingFeatureForMain = getFeature({
  id: "messaging-for-main",

  register: (di) => {
    autoRegister({
      di,
      targetModule: module,

      getRequireContexts: () => [require.context("./", true, /\.injectable\.(ts|tsx)$/)],
    });
  },
});
