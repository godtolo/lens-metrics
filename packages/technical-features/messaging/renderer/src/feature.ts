import { autoRegister } from "@ogre-tools/injectable-extension-for-auto-registration";
import { getFeature } from "@k8slens/feature-core";

export const messagingFeatureForRenderer = getFeature({
  id: "messaging-for-renderer",

  register: (di) => {
    autoRegister({
      di,
      targetModule: module,

      getRequireContexts: () => [require.context("./", true, /\.injectable\.(ts|tsx)$/)],
    });
  },
});
