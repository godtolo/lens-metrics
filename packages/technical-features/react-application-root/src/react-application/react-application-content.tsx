import { withInjectables } from "@ogre-tools/injectable-react";
import { computedInjectManyInjectable } from "@ogre-tools/injectable-extension-for-mobx";
import React from "react";
import {
  ReactApplicationChildren,
  reactApplicationChildrenInjectionToken,
} from "./react-application-children-injection-token";
import type { IComputedValue } from "mobx";
import { observer, Observer } from "mobx-react";

type Dependencies = { children: IComputedValue<ReactApplicationChildren[]> };

const NonInjectedContent = observer(({ children }: Dependencies) => (
  <>
    {children.get().map((child) => (
      <Observer key={child.id}>{() => (child.enabled.get() ? <child.Component /> : null)}</Observer>
    ))}
  </>
));

export const ReactApplicationContent = withInjectables<Dependencies>(
  NonInjectedContent,

  {
    getProps: (di) => ({
      children: di.inject(computedInjectManyInjectable)(reactApplicationChildrenInjectionToken),
    }),
  },
);
