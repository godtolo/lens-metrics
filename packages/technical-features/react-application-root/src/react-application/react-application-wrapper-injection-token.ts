import { getInjectionToken } from "@ogre-tools/injectable";
import type React from "react";

export type ReactApplicationWrapper = (Component: React.ComponentType) => React.ComponentType;

export const reactApplicationWrapperInjectionToken = getInjectionToken<ReactApplicationWrapper>({
  id: "react-application-wrapper-injection-token",
});
