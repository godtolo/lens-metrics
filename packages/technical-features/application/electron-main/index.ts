import { overrideSideEffectsWithFakes } from "./src/override-side-effects-with-fakes";

export { beforeElectronIsReadyInjectionToken } from "./src/start-application/timeslots/before-electron-is-ready-injection-token";
export { beforeAnythingInjectionToken } from "./src/start-application/timeslots/before-anything-injection-token";

export { applicationFeatureForElectronMain } from "./src/feature";

export const testUtils = { overrideSideEffectsWithFakes }
