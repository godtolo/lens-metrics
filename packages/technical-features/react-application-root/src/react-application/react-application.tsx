/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainerForInjection } from "@ogre-tools/injectable";
import { computedInjectManyInjectable } from "@ogre-tools/injectable-extension-for-mobx";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import { flow, identity } from "lodash/fp";
import { observer } from "mobx-react";
import React from "react";
import { reactApplicationWrapperInjectionToken } from "./react-application-wrapper-injection-token";

import { ReactApplicationContent } from "./react-application-content";

interface ReactApplicationProps {
  di: DiContainerForInjection;
}

export const ReactApplication = observer(({ di }: ReactApplicationProps) => {
  const computedInjectMany = di.inject(computedInjectManyInjectable);

  const wrappers = computedInjectMany(reactApplicationWrapperInjectionToken);

  const ContentWithWrappers = flow(identity, ...wrappers.get())(ReactApplicationContent);

  return (
    <DiContextProvider value={{ di }}>
      <ContentWithWrappers />
    </DiContextProvider>
  );
});
