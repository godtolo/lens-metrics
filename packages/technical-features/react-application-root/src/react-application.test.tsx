import { registerFeature } from "@k8slens/feature-core";
import { createContainer, DiContainer, getInjectable } from "@ogre-tools/injectable";
import { registerMobX } from "@ogre-tools/injectable-extension-for-mobx";
import { registerInjectableReact } from "@ogre-tools/injectable-react";
import { reactApplicationRootFeature } from "./feature";
import { runInAction, computed, observable, IObservableValue } from "mobx";
import { startApplicationInjectionToken } from "@k8slens/application";
import type { RenderResult } from "@testing-library/react";
import { render, act } from "@testing-library/react";
import renderInjectable from "./render-application/render.injectable";
import { reactApplicationChildrenInjectionToken } from "./react-application/react-application-children-injection-token";
import React from "react";
import { Discover, discoverFor } from "@k8slens/react-testing-library-discovery";
import { reactApplicationWrapperInjectionToken } from "./react-application/react-application-wrapper-injection-token";

const SomeChildren = () => <div data-some-children-test>Some children</div>;

describe("react-application", () => {
  let rendered: RenderResult;
  let di: DiContainer;
  let discover: Discover;

  beforeEach(async () => {
    di = createContainer("some-container");

    registerInjectableReact(di);

    registerMobX(di);

    runInAction(() => {
      registerFeature(di, reactApplicationRootFeature);
    });

    di.override(renderInjectable, () => (application) => {
      rendered = render(application);
    });

    const startApplication = di.inject(startApplicationInjectionToken);

    await startApplication();

    discover = discoverFor(() => rendered);
  });

  it("renders", () => {
    expect(rendered.baseElement).toMatchSnapshot();
  });

  describe("when children is registered and enabled", () => {
    let someObservable: IObservableValue<boolean>;

    beforeEach(() => {
      someObservable = observable.box(true);

      const someChildrenInjectable = getInjectable({
        id: "some-children",

        instantiate: () => ({
          id: "some-children",
          Component: SomeChildren,
          enabled: computed(() => someObservable.get()),
        }),

        injectionToken: reactApplicationChildrenInjectionToken,
      });

      runInAction(() => {
        di.register(someChildrenInjectable);
      });
    });

    it("renders", () => {
      expect(rendered.baseElement).toMatchSnapshot();
    });

    it("renders the children", () => {
      const { discovered } = discover.getSingleElement("some-children");

      expect(discovered).not.toBeNull();
    });

    describe("when wrapper is registered", () => {
      beforeEach(() => {
        const someWrapperInjectable = getInjectable({
          id: "some-wrapper",

          instantiate: () => (Component) => () =>
            (
              <div data-some-wrapper-test>
                <Component />
              </div>
            ),

          injectionToken: reactApplicationWrapperInjectionToken,
        });

        runInAction(() => {
          di.register(someWrapperInjectable);
        });
      });

      it("renders", () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it("renders the children inside the wrapper", () => {
        const { discovered } = discover
          .getSingleElement("some-wrapper")
          .getSingleElement("some-children");

        expect(discovered).not.toBeNull();
      });
    });

    describe("when children is enabled", () => {
      beforeEach(() => {
        act(() => {
          runInAction(() => {
            someObservable.set(false);
          });
        });
      });

      it("renders", () => {
        expect(rendered.baseElement).toMatchSnapshot();
      });

      it("does not render the children", () => {
        const { discovered } = discover.querySingleElement("some-children");

        expect(discovered).toBeNull();
      });
    });
  });
});
