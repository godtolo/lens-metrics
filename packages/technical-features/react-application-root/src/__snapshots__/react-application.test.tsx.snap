// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-application renders 1`] = `
<body>
  <div />
</body>
`;

exports[`react-application when children is registered and enabled renders 1`] = `
<body>
  <div>
    <div
      data-some-children-test="true"
    >
      Some children
    </div>
  </div>
</body>
`;

exports[`react-application when children is registered and enabled when children is enabled renders 1`] = `
<body>
  <div />
</body>
`;

exports[`react-application when children is registered and enabled when wrapper is registered renders 1`] = `
<body>
  <div>
    <div
      data-some-wrapper-test="true"
    >
      <div
        data-some-children-test="true"
      >
        Some children
      </div>
    </div>
  </div>
</body>
`;
